pkg/jobs/jobfrontier/frontier.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/span"
+	"github.com/cockroachdb/errors"
+)
+
+// ShardFunc is a function that determines which shard a span belongs to.
+// It returns the shard key suffix to append to the base info key.
+// If sharding is not needed, this can be nil.
+type ShardFunc func(span roachpb.Span) string
+
+// Unsharded is a nil ShardFunc that signals to not shard at all.
+var Unsharded ShardFunc = nil
+
+const frontierPrefix = "frontier/"
+const shardSep = "_"
+
+// Get loads a complete frontier from persistent storage and returns it.
+//
+// The returned frontier will contain all spans and their timestamps that were
+// previously stored via Store(). The spans are derived from the persisted data.
+func Get(
+	ctx context.Context,
+	txn isql.Txn,
+	jobID jobspb.JobID,
+	name string,
+) (span.Frontier, bool, error) {
+	infoStorage := jobs.InfoStorageForJob(txn, jobID)
+
+	// Read all persisted entries, both as entries and as plain spans; we need the
+	// latter form to construct the frontier and the former to advance it.
+	// TODO(dt): we could avoid duplicate allocation here if we added an API to
+	// construct frontier directly from entries.
+	var entries []frontierEntry
+	var spans []roachpb.Span
+
+	keyPrefix := frontierPrefix + name + shardSep
+
+	var found bool
+	if err := infoStorage.Iterate(ctx, keyPrefix, func(_ string, value []byte) error {
+		found = true
+		var r jobspb.ResolvedSpans
+		if err := protoutil.Unmarshal(value, &r); err != nil {
+			return err
+		}
+		for _, sp := range r.ResolvedSpans {
+			entries = append(entries, frontierEntry{Span: sp.Span, Timestamp: sp.Timestamp})
+			spans = append(spans, sp.Span)
+		}
+		return nil
+	}); err != nil || !found {
+		return nil, false, err
+	}
+
+	// Construct frontier to track the set of spans found and advance it to their
+	// persisted timestamps. This implies we perist zero-timestamp spans to keep
+	// the set of tracked spans even if they do not have progress.
+	frontier, err := span.MakeFrontier(spans...)
+	if err != nil {
+		return nil, false, err
+	}
+	for _, entry := range entries {
+		if _, err := frontier.Forward(entry.Span, entry.Timestamp); err != nil {
+			return nil, false, err
+		}
+	}
+
+	return frontier, true, nil
+}
+
+// Store persists a frontier's current state to storage.
+//
+// All span entries in the frontier and their current timestamps will be
+// persisted. Any previously stored frontier data under the same key(s) will
+// be replaced.
+//
+// The shardFunc, if non-nil, will be used to bucket spans into 'shards' that
+// are stored in separate info keys under the passed name prefix. If nil, all
+// spans are stored in one key, using the shard name "full".
+//
+// InfoStorage keys are prefixed with "frontier/", and all shards, including the
+// single, implicit "full" shard for unsharded, are under the passed name, so a
+// full key would be "frontier/{name}_full".
+func Store(
```

Comment:
An alternative API to consider:

```
func Store(ctx context.Context, txt isql.Txn, jobID jobspb.JobID, name string, shard string, frontier span.Frontier) error
```

This is more flexible and less prescriptive. For example, it would allow for each processor in a distflow to checkpoint its own frontier.

We could offer a utility to split a frontier into shards. E.g.

```
for name, shard := range frontier.Split(shardFunc) {

}
```

Reply:
> For example, it would allow for each processor in a distflow to checkpoint its own frontier.

they can just do that by passing a processor specific name though?

the point of shardfunc in the signature is to nudge the caller to consider sharding before just calling Store() on a huge frontier